libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,207 @@
+			if (!m_graph->edges.count(funcDef))
+				visitCallable(funcDef);
+	}
+
+	// Add all CreationDispatch calls to the RuntimeDispatch as well
+	for (auto node: m_graph->edges[SpecialNode::CreationDispatch])
+		add(SpecialNode::InternalDispatch, node);
+
+	// Add all external functions to the RuntimeDispatch
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		add(SpecialNode::ExternalDispatch, &functionType->declaration());
+	}
+
+	if (_contract.fallbackFunction())
+		add(SpecialNode::ExternalDispatch, _contract.fallbackFunction());
+
+	if (_contract.receiveFunction())
+		add(SpecialNode::ExternalDispatch, _contract.receiveFunction());
+
+	m_contract = nullptr;
+	solAssert(!m_currentNode.has_value(), "Current node not properly reset.");
+
+	return m_graph;
+}
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		solAssert(*_identifier.annotation().requiredLookup == VirtualLookup::Virtual, "");
+
+		auto funType = dynamic_cast<FunctionType const*>(_identifier.annotation().type);
+
+		// For events kind() == Event, so we have an extra check here
+		if (funType && funType->kind() == FunctionType::Kind::Internal)
+		{
+			processFunction(callable->resolveVirtual(*m_contract), _identifier.annotation());
+
+			solAssert(m_currentNode.has_value(), "");
+		}
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+		m_graph->createdContracts.emplace(&contractType->contractDefinition());
+
+	return true;
+}
+
+void FunctionCallGraphBuilder::endVisit(MemberAccess const& _memberAccess)
+{
+	auto functionType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type);
+	auto functionDef = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration);
+	if (!functionType || !functionDef || functionType->kind() != FunctionType::Kind::Internal)
+		return;
+
+	// Super functions
+	if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+		if (ContractType const* type = dynamic_cast<ContractType const*>(_memberAccess.expression().annotation().type))
+		{
+			solAssert(type->isSuper(), "");
+			functionDef = &functionDef->resolveVirtual(*m_contract, type->contractDefinition().superContract(*m_contract));
+		}
+
+	processFunction(*functionDef, _memberAccess.annotation());
+	return;
+}
+
+void FunctionCallGraphBuilder::endVisit(FunctionCall const& _functionCall)
+{
+	auto* functionType = dynamic_cast<FunctionType const*>(_functionCall.expression().annotation().type);
+
+	if (
+		functionType &&
+		functionType->kind() == FunctionType::Kind::Internal &&
+		!functionType->hasDeclaration()
+	)
+		add(m_currentDispatch, &_functionCall);
```

Comment:
Shouldn't this instead cause an edge from the current function to the dispatch?

Reply:
When we visit the identifier and it is not called immediately, it should cause an edge from the dispatch to that function (the function is added to the dispatch in the code generator).

Reply:
I feel your two statements here are in direct contradiction to each other. Either it should be dispatch -> uncalled function or uncalled function -> dispatch... from what I originally understood it is supposed to be the former.

Reply:
There should be an edge from the identifier (the function it is in) to the dispatch and from the dispatch to the function that is referenced by the indentifier. If the function is called right away (-> annotation), then the edge can go directly to the function without going through the dispatch.

Reply:
Done. It now works as described by @chriseth.